### PR TITLE
Support Cast on Result<T>

### PR DESCRIPTION
**Pull request by Corniel**
To support the cast between `Result<T>` and `Result<TOut>` where `TOut` is a subtype of `T`:

``` C#
var result = Result.For<T>(...);
var casted = result.Cast<TOut>();
```

### CODE REVIEW

File: specs/Qowaiv.Validation.Specs/Abstractions/Result_specs.cs [hunk 1]
```diff
@@ -99,14 +100,37 @@ public class Casting
...
     [Test]
     public void Explicit_from_Result_of_T_to_T_is_supported()
     {
         var result = Result.For(666);
-        Assert.That((int)result, Is.EqualTo(666));
+        result.Should().BeValid().Value.Should().Be(666);
+    }
+
+    [Test]
+    public void With_method_to_Result_of_valid__wiht_subtype_of_T_is_supported()
```
Comment (renzo-baasdam) [inline specs/Qowaiv.Validation.Specs/Abstractions/Result_specs.cs:114]: __wiht -> _with
Also note the double _

File: specs/Qowaiv.Validation.Specs/Abstractions/Result_specs.cs [hunk 2]
```diff
@@ -99,14 +100,37 @@ public class Casting
...
+
+
+    [Test]
+    public void With_method_to_Result_of_invalid_with_subtype_of_T_is_supported()
+    {
+        Result.WithMessages<XDocument>(Error1).Cast<XNode>().Should().BeInvalid();
+    }
+
+    [Test]
+    public void With_method_is_not_supported_TOut_not_bein_subtype_of_TModel()
```
Comment (renzo-baasdam) [inline specs/Qowaiv.Validation.Specs/Abstractions/Result_specs.cs:129]: bein -> being
Also the is_supported is on the end on the previous tests. Also is_not_supported here at the end for consistency?

**Review by renzo-baasdam — APPROVED** [review]

**Review by RensReus — APPROVED** [review]

**Review by Sjaaky — APPROVED** [review]
lgtm

**Comment by Corniel** [comment]
> I'm not an expert, but from the looks of it, the ['out' keyword](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-generic-modifier) should work in this case.
> 
> Just replace `public sealed class Result<TModel> : Result` with `public sealed class Result<out TModel> : Result`
> 
> -- edit: This only works on interfaces and delegates.

Unfortunately. :(

**Pull request merged**